src/drivers/ins/microstrain/MicroStrain.cpp

```diff
@@ -0,0 +1,623 @@
+
+MicroStrain::MicroStrain(const char *uart_port, int32_t rot) :
+	ModuleParams(nullptr),
+	ScheduledWorkItem(MODULE_NAME, px4::wq_configurations::test1)
+{
+	/* store port name */
+	memset(_port, '\0', 20);
+	int max_len = math::min((unsigned int)20, strlen(uart_port));
+	strncpy(_port, uart_port, max_len);
+	/* enforce null termination */
+	_port[19] = '\0';
+
+	rotation = static_cast<Rotation>(rot);
+	ms_schedule_rate_us = (1000000) / (2 * _param_ms_imu_rate_hz.get());
+
+	device::Device::DeviceId device_id{};
+	device_id.devid_s.devtype = DRV_INS_DEVTYPE_MS;
+	device_id.devid_s.bus_type = device::Device::DeviceBusType_SERIAL;
+	device_id.devid_s.bus = 2;
+	dev_id = device_id.devid;
+
+
+	// Default to ROTATION_NONE
+	_px4_accel.set_device_id(dev_id);
+	_px4_gyro.set_device_id(dev_id);
+	_px4_mag.set_device_id(dev_id);
+
+	// Set the default values for the baro (which may not change)
+	_sensor_baro.device_id = dev_id;
+	_sensor_baro.pressure = 0;
+	_sensor_baro.temperature = 0;
+	_sensor_baro.error_count = 0;
+}
+
+MicroStrain::~MicroStrain()
+{
+	if (device_uart.is_open()) {
+		device_uart.uart_close();
+	}
+
+	PX4_INFO("Destructor");
+	_sensor_baro_pub.unadvertise();
+
+	perf_free(_loop_perf);
+	perf_free(_loop_interval_perf);
+}
+
+bool mip_interface_user_recv_from_device(mip_interface *device, uint8_t *buffer, size_t max_length,
+		timeout_type wait_time,
+		size_t *out_length, timestamp_type *timestamp_out)
+{
+	(void)device;
+
+	*timestamp_out = hrt_absolute_time();
+
+	int res = device_uart.uart_read(buffer, max_length);
+
+	if (res == -1 && errno != EAGAIN) {
+		PX4_DEBUG("RX 1 %d(%d)", res, max_length);
+		*out_length = 0;
+		return false;
+	}
+
+	if (res >= 0) {
+		*out_length = res;
+		PX4_DEBUG("RX 2 %d(%d)", *out_length, max_length);
+	}
+
+	PX4_DEBUG("RX 3 %d(%d)", *out_length, max_length);
+
+	return true;
+}
+
+bool mip_interface_user_send_to_device(mip_interface *device, const uint8_t *data, size_t length)
+{
+
+	int res = device_uart.uart_write(const_cast<uint8_t *>(data), length);
+
+	if (res >= 0) {
+		return true;
+	}
+
+	return false;
+
+}
+
+int MicroStrain::connect_at_baud(int32_t baud)
+{
+	if (device_uart.is_open()) {
+		if (device_uart.uart_set_baud(baud) == PX4_ERROR) {
+			PX4_INFO(" - Failed to set UART %" PRIu32 " baud", baud);
+		}
+
+	} else if (device_uart.uart_open(_port, baud) == PX4_ERROR) {
+		PX4_INFO(" - Failed to open UART");
+		PX4_ERR("ERROR: Could not open device port!");
+		return PX4_ERROR;
+	}
+
+	PX4_INFO("Serial Port %s with baud of %" PRIu32 " baud", (device_uart.is_open() ? "CONNECTED" : "NOT CONNECTED"), baud);
+
+	// Re-init the interface with the correct timeouts
+	mip_interface_init(&device, parse_buffer, sizeof(parse_buffer), mip_timeout_from_baudrate(baud) * 1_ms, 250_ms,
+			   &mip_interface_user_send_to_device, &mip_interface_user_recv_from_device, &mip_interface_default_update, NULL);
+
+	PX4_INFO("mip_base_ping");
+
+	if (mip_base_ping(&device) != MIP_ACK_OK) {
```

Comment:
If the device is setup to stream on startup, you may miss the response from a ping on lower baudrates even if you repeat it. Although this is a configuration problem, it would be good to let users know.

To deal with this in the driver, the first thing I do after opening the port is to try to set it to idle 3 times. If I succeed by the third time, I know I am at the right baudrate, and the device is valid. [This](https://github.com/LORD-MicroStrain/microstrain_inertial_driver_common/blob/main/src/utils/mip/ros_mip_device_main.cpp#L45-L80) is the relevant code in the driver